texar/modules/regressors/xlnet_regressor.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+XLNet Regressors.
+"""
+
+from typing import Any, Dict, Iterable, List, Optional
+
+import itertools
+
+import torch
+import torch.nn as nn
+
+from texar.modules.regressors.regressor_base import RegressorBase
+from texar.modules.encoders.xlnet_encoder import XLNetEncoder
+from texar.utils import utils
+
+
+__all__ = [
+    "XLNetRegressor",
+]
+
+
+class XLNetRegressor(RegressorBase):
```

Comment:
Do we really need a separate regressor here? I implemented a regressor because both it and the classifier could share a common base class `XLNetSummary`. Now, with this middle layer is gone, there's a lot of duplicated code among the two classes.
Is it possible to do regression using classifiers in Texar (e.g. by setting `num_classes` to zero)? If it is, I think we can remove this class.

Reply:
I think `XLNetRegressor` is necessary here. I added some new logic to `XLNetClassifier` following the logic of `BertClassifier`. If `num_classes <= 0` in `XLNetClassifer`, we do not append the dense layer, and the number of classes is assumed to be the final dense layer size of the encoder.

Reply:
It is true that the code structure of `XLNetRegressor` and `XLNetClassifier` are very similar. But this might be a cleaner way to present our modules. Please let me know your thought.

Reply:
I'd suggest keeping regressor modules separate from classifiers. But we shall reduce the code duplication, e.g., by sharing the code portion in `__init__` (encoder-to-logits) possibly by wrapping it as a function.

Reply:
Sharing `__init__` would require `XLNetRegressor` to inherit from `XLNetClassifier`, which I think is kind of weird. Merging the regressor interface with classifiers would mean that all existing classifiers (Conv, BERT, GPT2) could be used as regressors out-of-the-box.

Reply:
Using a module named `*Classifier` for regression is a bit weird. Can we do sth like:
```
class XLNetClassifier:
    def __init__(...):
        ... = XLNetEncoder(...)
        a_shared_function_building_output_layers(...)
        ... classifier_stuff ...

class XLNetRegressor:
    def __init__(...):
        ... = XLNetEncoder(...)
        a_shared_function_building_output_layers(...)
        ... regressor_stuff ...
```

Or, can we have sth like the previous `XLNetSummary` intermediate class, which is not exposed to users?

Can we allow users to config the output layers (e.g., number of layers, layer sizes, etc)? Is it necessary to allow it?

Reply:
We currently only support one hidden-to-logits layer, and it is pretty straightforward to support multi-layer feature. But I am not sure if this is a helpful feature. Maybe we can make it a future enhancement?

Reply:
I don't really think it's necessary to add this feature. These classifiers are simply wrappers anyway, if the user needs this they can easily write their own classifiers.